ext/kv/sqlite.rs

```diff
@@ -132,44 +133,91 @@ impl<P: SqliteDbHandlerPermissions> DatabaseHandler for SqliteDbHandler<P> {
-        rusqlite::Connection::open_with_flags(path, flags)?
-      }
-      (None, Some(path)) => {
-        std::fs::create_dir_all(path)?;
-        let path = path.join("kv.sqlite3");
-        rusqlite::Connection::open(&path)?
-      }
-    };
-
-    conn.pragma_update(None, "journal_mode", "wal")?;
-    conn.execute(STATEMENT_CREATE_MIGRATION_TABLE, [])?;
-
-    let current_version: usize = conn
-      .query_row(
-        "select version from migration_state where k = 0",
-        [],
-        |row| row.get(0),
-      )
-      .optional()?
-      .unwrap_or(0);
-
-    for (i, migration) in MIGRATIONS.iter().enumerate() {
-      let version = i + 1;
-      if version > current_version {
-        conn.execute_batch(migration)?;
-        conn.execute(
-          "replace into migration_state (k, version) values(?, ?)",
-          [&0, &version],
-        )?;
       }
     }
 
-    Ok(SqliteDb(RefCell::new(conn)))
+    let default_storage_dir = self.default_storage_dir.clone();
+    let conn = spawn_blocking(move || {
+      let conn = match (path.as_deref(), &default_storage_dir) {
+        (Some(":memory:"), _) | (None, None) => {
+          rusqlite::Connection::open_in_memory()?
+        }
+        (Some(path), _) => {
+          let flags =
+            OpenFlags::default().difference(OpenFlags::SQLITE_OPEN_URI);
+          rusqlite::Connection::open_with_flags(path, flags)?
+        }
+        (None, Some(path)) => {
+          std::fs::create_dir_all(path)?;
+          let path = path.join("kv.sqlite3");
+          rusqlite::Connection::open(&path)?
+        }
+      };
+
+      conn.pragma_update(None, "journal_mode", "wal")?;
+      conn.execute(STATEMENT_CREATE_MIGRATION_TABLE, [])?;
+
+      let current_version: usize = conn
+        .query_row(
+          "select version from migration_state where k = 0",
+          [],
+          |row| row.get(0),
+        )
+        .optional()?
+        .unwrap_or(0);
+
+      for (i, migration) in MIGRATIONS.iter().enumerate() {
+        let version = i + 1;
+        if version > current_version {
+          conn.execute_batch(migration)?;
+          conn.execute(
+            "replace into migration_state (k, version) values(?, ?)",
+            [&0, &version],
+          )?;
+        }
+      }
+
+      Ok::<_, AnyError>(conn)
+    })
+    .await
+    .unwrap()?;
+
+    Ok(SqliteDb(Rc::new(AsyncRefCell::new(Cell::new(Some(conn))))))
   }
 }
 
-pub struct SqliteDb(RefCell<rusqlite::Connection>);
+pub struct SqliteDb(Rc<AsyncRefCell<Cell<Option<rusqlite::Connection>>>>);
+
+impl SqliteDb {
+  async fn run_tx<F, R>(&self, f: F) -> Result<R, AnyError>
+  where
+    F: (FnOnce(rusqlite::Transaction<'_>) -> Result<R, AnyError>)
+      + Send
+      + 'static,
+    R: Send + 'static,
+  {
+    // We need exclusive access to the db to run a transaction. Wait until we can borrow_mut the db.
+    let cell = self.0.borrow_mut().await;
```

Comment:
We're still using a single connection here. In the future we can add a conn pool if needed.

Reply:
Note that before adding pooling we need to properly handle SQLite lock errors